apis/v1alpha2/shared_types.go

```diff
@@ -383,3 +383,58 @@ type AnnotationKey string
 // +kubebuilder:validation:MinLength=0
 // +kubebuilder:validation:MaxLength=4096
 type AnnotationValue string
+
+// TrafficMatches defines AddressMatch rules for inbound traffic according to
+// source and/or destination address of a packet or connection.
+type TrafficMatches struct {
+	// SourceAddresses indicates the originating (source) network
+	// addresses which are valid for routing traffic.
+	//
+	// Support: Core
+	SourceAddresses []AddressMatch `json:"sourceAddresses"`
+
+	// DestinationAddresses indicates the destination network addresses
+	// which are valid for routing traffic.
+	//
+	// Support: Core
+	DestinationAddresses []AddressMatch `json:"destinationAddresses"`
```

Comment:
If we followed the API pattern from HTTPRoute then this struct could have the following fields:

```go
SourceAddress *Address
SourceNetwork *NetworkAddress
DestinationAddress *Address
DestinationNetwork *NetworkAddress
```

The conditions in this struct would be AND'ed. The matches slice would be OR'ed.

Reply:
It seems like your suggestion here proposes two things:

1. adding the `NetworkAddress` type
2. removing the `AddressMatch` type

Removing the `AddressMatch` type in favor of flat `Address` removes the forward flexibility that was intended by the `type+value` definition of an `Address` which followed the previous pattern established by `GatewayAddress` for how a network address can be defined. In your proposed model would you expect those new types to be added horizontally alongside `SourceAddress`, `DesitinationAddress`, e.t.c. as they come up? Just trying to make sure I'm understanding correctly.

If I'm understanding your suggestion correctly It seems that `NetworkAddress` is an alternate from the previously suggested (but noted as a consideration for later, see the alternatives section of the GEP) permutation of the `Address` type (such as a `CIDR` type) for match inclusion and solves the problem by adding horizontal options. The existing way solves the problem by nesting the permutations via the `type+value` option which intentionally followed the design pattern already present in the `GatewayAddress` type.

I'm not opposed to the change but It is not immediately clear to me that there is a benefit, could you help me to better understand your reasoning? Also given that this seems like a significant divergence from the original GEP I'm inclined to say it should be a follow up so that we can discuss that reasoning in isolation and forum at length and not overload this PR with that conversation, but I'm flexible and if most people prefer to sort this out now I will go along with it.

Reply:
I agree with @shaneutt here, I'm not opposed to a change, but don't see a clear benefit. Some additional explanation could be helpful. If we were to go ahead with a change like that, maybe the best approach would be a separate PR to propose changes to the GEP itself.

Reply:
Sorry, the main gist of this suggestion is that we no longer need slices here. Since the outer struct has a slice (with OR semantics), this struct must have AND semantics, so the slice becomes moot (you can only use 1 element in almost all cases).

The proposal of a `NetworkAddress` type is to clarify the distinction between individual addresses and CIDR addresses, which (based on the discussion is not crystal clear). I think that it makes sense in this API to be able to configure network ranges, and that is probably also makes sense to have an explicit type for that with it's own nomenclature and validation.

Maybe it makes sense to mirror the existing `AddressType` to `NetworkType` so you can have a CIDR network, or a named network if it makes sense. I think that's pretty consistent with HTTPRoute and you wouldn't need to duplicate the existing gateway address.

Still thinking out loud here, but if you can match traffic on a destination CIDR network, that would imply that you have make a gateway listen on a network address, and I don't think there's a way to express that.

Happy to discuss further on slack or zoom :)

Reply:
> Sorry, the main gist of this suggestion is that we no longer need slices here. Since the outer struct has a slice (with OR semantics), this struct must have AND semantics, so the slice becomes moot (you can only use 1 element in almost all cases).

In the GEP we discussed (and originally had) horizontal options at this level for destination vs source, and address vs port. To me it still seems useful to have this distinction for forward growth, e.g. direction of traffic and type of identifier (address, port, e.t.c.) would be horizontal here and then for each of those identifiers there would be a nested type to further configure those individual matches.

> The proposal of a NetworkAddress type is to clarify the distinction between individual addresses and CIDR addresses, which (based on the discussion is not crystal clear).
> I think that it makes sense in this API to be able to configure network ranges, and that is probably also makes sense to have an explicit type for that with it's own nomenclature and validation.
> Maybe it makes sense to mirror the existing AddressType to NetworkType so you can have a CIDR network, or a named network if it makes sense. I think that's pretty consistent with HTTPRoute and you wouldn't need to duplicate the existing gateway address.
> Still thinking out loud here, but if you can match traffic on a destination CIDR network, that would imply that you have make a gateway listen on a network address, and I don't think there's a way to express that.

CIDR was noted but specifically left out of the first iteration of the GEP (https://github.com/kubernetes-sigs/gateway-api/blob/master/site-src/geps/gep-735.md#cidr-addresstype). I think we should be doing follow up GEP work to add CIDR matching rather than trying to solve for that problem in a PR that is only meant to implement the state of the GEP so far.

Reply:
Deferring CIDR network address types makes sense to me.

> In the GEP we discussed (and originally had) horizontal options at this level for destination vs source, and address vs port. 

Sorry, I'm not familiar with what you mean by "horizontal options", so not sure whether we are in agreement or not :)

Reply:
Attributes on the same struct, vs sub-structs.

Reply:
OK. All I think that we need here now it to switch to 
```Go
	SourceAddress *AddressMatch `json:"sourceAddress"`
	DestinationAddress *AddressMatch `json:"destinationAddress"`
```

and document that these fields are AND match conditions. We should crib some of the language fro HTTPRouteMatch to document the semantics.

Reply:
I cherry picked 563873c25289a4abe0bb9f1fb09060bd16c6b4e5 into this branch from your fork (as we discussed separately in slack). I'm OK with the suggested change, so I will consider this resolved but if anyone feels there's more to be discussed here please do feel free to unresolve this comment :+1: